File: backend/compliance.py
"""
compliance.py — Did the standard land everywhere? (#543)

`config_push.preview` can already say, of one device, which lines of a
block are present and which are missing — without sending anything, from
the configuration ShellMate already holds. Nothing asked that question of
a *group*, and "these fourteen switches are missing the new AAA lines" is
the change-management question at two hundred sites.

It needs no login. The scheduled backups keep every device's latest
snapshot fresh, so this reads what is already stored and answers
immediately. That is the whole appeal, and it is also the thing that makes
the answer subtle:

**Snapshot age is part of the answer, not metadata about it.** "Compliant"
against a capture from six weeks ago is a statement about six weeks ago. A
row that reports compliance without saying how old the evidence is invites
exactly the wrong conclusion, so the age travels with every verdict and a
device with no snapshot at all is its own state rather than being rounded
to either compliant or not.

**Three states, not two.** `compliant`, `missing`, and `never-captured`.
Folding the third into "not compliant" would send somebody to fix a device
that may be perfectly configured; folding it into "compliant" is worse.

**The limit is stated, not hidden.** Lines are matched as a set with
indentation stripped, so section context is ignored — `description uplink`
under the wrong interface counts as present. That is right for the flat
blocks this is for (AAA, logging, NTP, SNMP) and wrong for anything whose
meaning depends on the line above it. Every result carries `limits` saying
so, and the panel prints it. A check that overstates what it verified is
worse than no check.
"""

import logging
import time
from typing import Any

logger = logging.getLogger(__name__)

#: How old a snapshot can be before the verdict is flagged as stale. A
#: nightly backup makes everything a day old at worst; a week means the
#: schedule is not running, which is itself the finding.
STALE_AFTER_DAYS = 7

LIMITS = (
    "Lines are compared as a set, with indentation ignored, so a line that "
    "appears under a different parent still counts as present. This suits "
    "flat blocks — AAA, logging, NTP, SNMP — and not configuration whose "
    "meaning depends on the section it sits in."
)


def _platform_of(profile: dict) -> str:
    """
    What ShellMate believes this device is.

    The remembered observation first, then whatever was set by hand. A
    platform nobody has established is "", which callers treat as "any
    snippet applies" rather than as a reason to skip the device.
    """
    return (str(profile.get("last_seen_platform") or "").strip()
            or str(profile.get("platform") or "").strip())


def _snippet_lines(snippet: Any) -> list[str]:
    """The commands of a snippet, however it arrived."""
    if snippet is None:
        return []
    if isinstance(snippet, dict):
        commands = snippet.get("commands") or []
    else:
        commands = getattr(snippet, "commands", None) or []
    return [str(c) for c in commands if str(c).strip()]


def check_device(profile: dict, snippet: Any, latest_snapshot,
                 must_not_have: Any = None) -> dict:
    """
    One device's verdict.

    Args:
        profile:         The connection profile.
        snippet:         The golden block — a Snippet, a dict, or None.
        latest_snapshot: ``hostname -> snapshot row or None``. Injected so a
                         test needs no database and the caller can read the
                         whole group in one pass if it ever wants to.
        must_not_have:   An optional block that should *not* be present.

    Returns:
        A row for the table: name, hostname, platform, state, the missing
        and unexpected lines, and how old the evidence is.
    """
    from backend import config_push

    name = (profile.get("name") or profile.get("hostname")
            or profile.get("id", "?"))
    hostname = (str(profile.get("hostname") or "").strip()
                or str(profile.get("name") or "").strip())
    platform = _platform_of(profile)

    row: dict = {
        "name": name, "hostname": hostname, "platform": platform,
        "state": "never-captured", "missing": [], "unexpected": [],
        "present": 0, "captured_at": None, "age_days": None, "stale": False,
    }

    if not hostname:
        row["state"] = "no-device-name"
        return row

    snapshot = latest_snapshot(hostname)
    if not snapshot or not (snapshot.get("content") or "").strip():
        # Deliberately not "not compliant". A device nobody has captured
        # may be perfectly configured, and sending somebody to fix it is
        # the wrong instruction; so is calling it compliant.
        return row

    content = snapshot.get("content") or ""
    captured_at = snapshot.get("captured_at")
    row["captured_at"] = captured_at
    if captured_at:
        age = max(0.0, (time.time() - float(captured_at)) / 86400.0)
        row["age_days"] = round(age, 1)
        row["stale"] = age > STALE_AFTER_DAYS

    wanted = _snippet_lines(snippet)
    if not wanted:
        row["state"] = "no-snippet"
        return row

    result = config_push.check(content, wanted, platform)
    row["missing"] = [line["text"] for line in result["lines"]
                      if line["status"] == "add"]
    row["present"] = result["counts"]["present"]

    forbidden = _snippet_lines(must_not_have)
    if forbidden:
        # The same call. Anything that comes back present is a line that
        # should not be there — a second mode would be a second thing to
        # keep correct for no gain.
        against = config_push.check(content, forbidden, platform)
        row["unexpected"] = [line["text"] for line in against["lines"]
                             if line["status"] == "present"]

    row["state"] = ("compliant" if not row["missing"] and not row["unexpected"]
                    else "missing")
    return row


def check_group(key: str, profiles: list[dict], snippet_for, latest_snapshot,
                must_not_have_for=None) -> dict:
    """
    Ask the same question of every device in a group.

    Args:
        key:               The group.
        profiles:          Its connections, nested ones included.
        snippet_for:       ``platform -> snippet``. A callable rather than a
                           single snippet because a mixed group needs one
                           block per platform — the AAA lines for IOS are
                           not the AAA lines for NX-OS, and running the IOS
                           block against a firewall would report every line
                           missing and be worse than useless.
        latest_snapshot:   ``hostname -> snapshot row or None``.
        must_not_have_for: ``platform -> snippet``, or None.

    Returns:
        The table, its counts, when it ran, and the limits it was subject
        to. Never raises: one unreadable profile must not lose the other
        199 answers.
    """
    started = time.time()
    rows: list[dict] = []

    for profile in profiles:
        try:
            platform = _platform_of(profile)
            row = check_device(
                profile,
                snippet_for(platform) if snippet_for else None,
                latest_snapshot,
                must_not_have_for(platform) if must_not_have_for else None)
        except Exception as exc:                          # pragma: no cover
            logger.warning("Compliance check failed for %r: %s",
                           profile.get("name"), exc)
            row = {"name": profile.get("name") or "?", "hostname": "",
                   "platform": "", "state": "error", "missing": [],
                   "unexpected": [], "present": 0, "captured_at": None,
                   "age_days": None, "stale": False, "why": str(exc)[:200]}
        rows.append(row)

    # Worst first: a table of two hundred devices is read from the top, and
    # the compliant ones are the rows nobody needs to see.
    order = {"missing": 0, "never-captured": 1, "no-device-name": 2,
             "no-snippet": 3, "error": 4, "compliant": 5}
    rows.sort(key=lambda r: (order.get(r["state"], 9), r["name"].lower()))

    counts: dict[str, int] = {}
    for row in rows:
        counts[row["state"]] = counts.get(row["state"], 0) + 1

    return {
        "group": key,
        "at": started,
        "took_s": round(time.time() - started, 1),
        "devices": rows,
        "counts": counts,
        "checked": len(rows),
        # Carried with the result rather than left to the panel to
        # remember: a result forwarded, exported or read from the digest
        # has to arrive with the caveat attached to it.
        "limits": LIMITS,
        "stale_after_days": STALE_AFTER_DAYS,
    }


def summary_line(report: dict) -> str:
    """
    One sentence, for the digest and the group card.

    Names the two numbers somebody acts on and stays quiet about the rest.
    "14 of 60 missing lines" is a morning's work; "46 compliant" is not
    news, and a summary that leads with it gets skimmed past.
    """
    counts = report.get("counts") or {}
    missing = counts.get("missing", 0)
    never = counts.get("never-captured", 0)
    checked = report.get("checked", 0)

    if not checked:
        return "Nothing to check."
    if not missing and not never:
        return f"All {checked} device(s) have the whole block."

    bits = []
    if missing:
        bits.append(f"{missing} of {checked} missing lines")
    if never:
        bits.append(f"{never} never captured")
    return ", ".join(bits) + "."

File: frontend/js/compliance.js
/**
 * compliance.js — Did the standard land everywhere? (#543)
 *
 * A group, a golden snippet, and a table saying which devices are missing
 * which lines. It reads the snapshots the nightly backups already stored,
 * so it needs no login and answers at once.
 *
 * **Three states get three treatments, and that is the point.** A device
 * nobody has captured is not "not compliant" — it may be perfectly
 * configured, and sending somebody to fix it is the wrong instruction. Nor
 * is it compliant, which would report a device no one has looked at as
 * verified. It gets its own row style and its own count.
 *
 * **The caveat is printed, not remembered.** Lines are compared as a set
 * with indentation stripped, so a line under the wrong parent counts as
 * present. That is right for the flat blocks this is for and wrong
 * otherwise, and a check that overstates what it verified is worse than no
 * check. The server sends the sentence; this prints it above the table
 * rather than holding its own copy that could drift from the matching.
 */
(function () {
  'use strict';

  let overlay, current = null;

  function el(tag, attrs, children) {
    const node = document.createElement(tag);
    Object.entries(attrs || {}).forEach(([k, v]) => {
      if (k === 'text') node.textContent = v;
      else if (k === 'class') node.className = v;
      else if (k === 'onclick') node.addEventListener('click', v);
      else node.setAttribute(k, v);
    });
    (Array.isArray(children) ? children : children ? [children] : [])
      .forEach(c => node.appendChild(typeof c === 'string'
        ? document.createTextNode(c) : c));
    return node;
  }

  function close() {
    if (overlay) overlay.classList.add('hidden');
  }

  function init() {
    overlay = document.getElementById('compliance-overlay');
    if (!overlay) return;
    document.getElementById('compliance-close')
      .addEventListener('click', close);
    overlay.addEventListener('click', (e) => {
      if (e.target === overlay) close();
    });
    document.addEventListener('keydown', (e) => {
      if (e.key === 'Escape' && !overlay.classList.contains('hidden')) close();
    });
  }

  /**
   * Ask which snippet to check against, then run it.
   *
   * The library is fetched first so the picker lists real snippets rather
   * than asking somebody to remember an id. A snippet that writes is
   * offered like any other — this never sends anything, and refusing to
   * *compare* against a block because applying it would change something
   * would be a rule about the wrong verb.
   */
  async function open(group) {
    if (!group || !group.key) return;

    let library = [];
    try {
      const res = await fetch('/api/snippets');
      const data = await res.json();
      library = Array.isArray(data) ? data : (data.snippets || []);
    } catch (_) { library = []; }

    if (!library.length) {
      if (window.shellmateAlerts) window.shellmateAlerts.notify({
        global: true, severity: 'warning', icon: 'error',
        title: 'Nothing to check against',
        body: 'Save a snippet holding the lines every device should have, '
            + 'then check the group against it.' });
      return;
    }

    const options = library.map(s => ({
      value: s.id,
      label: s.platform ? `${s.name} (${s.platform})` : s.name,
    }));

    const answer = await window.shellmateDialog.form({
      title: `Check ${group.name} against a snippet`,
      body: 'Every device in the group is compared with the lines in the '
          + 'snippet, using the configuration ShellMate already has stored. '
          + 'Nothing is sent and nothing is logged into.',
      fields: [
        { name: 'snippet', label: 'Every device should have', type: 'select',
          options },
        { name: 'forbidden', label: 'And should not have (optional)',
          type: 'select',
          options: [{ value: '', label: 'Nothing in particular' }, ...options] },
        // Off by default. Somebody asking a one-off question about a site
        // has not asked for a nightly one, and a check that starts
        // recurring because it was run once is a surprise.
        { name: 'every_night', type: 'checkbox', value: false,
          label: 'Check again after every scheduled backup, and report it '
               + 'in the morning digest' },
      ],
      confirmLabel: 'Check',
    });
    if (!answer || !answer.snippet) return;

    await run(group, answer.snippet, answer.forbidden || '',
              !!answer.every_night);
  }

  async function run(group, snippetId, forbiddenId, everyNight) {
    let report;
    try {
      const res = await fetch('/api/groups/' + encodeURIComponent(group.key)
                              + '/compliance', {
        method: 'POST', headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ snippet_id: snippetId,
                               must_not_have_id: forbiddenId || '',
                               every_night: !!everyNight }),
      });
      report = await res.json().catch(() => ({}));
      if (!res.ok) throw new Error(report.detail || `HTTP ${res.status}`);
    } catch (e) {
      if (window.shellmateAlerts) window.shellmateAlerts.notify({
        global: true, severity: 'warning', icon: 'error',
        title: 'The check could not be run', body: String(e.message || e) });
      return;
    }
    show(group, report);
  }

  function show(group, report) {
    if (!overlay) return;
    current = report;

    document.getElementById('compliance-title').textContent =
      `${group.name} — compliance`;
    document.getElementById('compliance-summary').textContent =
      report.summary || '';

    // Printed, not remembered. A caveat the panel owns is one a forwarded
    // or exported result loses.
    const limits = document.getElementById('compliance-limits');
    limits.textContent = report.limits || '';
    limits.classList.toggle('hidden', !report.limits);

    const body = document.getElementById('compliance-body');
    body.innerHTML = '';
    (report.devices || []).forEach(row => body.appendChild(deviceRow(row)));

    overlay.classList.remove('hidden');
  }

  const STATE_LABEL = {
    compliant: 'Has every line',
    missing: 'Missing lines',
    'never-captured': 'Never captured',
    'no-device-name': 'No device name',
    'no-snippet': 'No block for this platform',
    error: 'Could not be checked',
  };

  function deviceRow(row) {
    const card = el('div', { class: `compliance-row compliance-${row.state}` });

    const head = el('div', { class: 'compliance-head' }, [
      el('span', { class: 'compliance-name', text: row.name }),
      el('span', { class: `compliance-state compliance-state-${row.state}`,
                   text: STATE_LABEL[row.state] || row.state }),
    ]);

    // The age of the evidence sits on the verdict, not beside it: a row
    // that says "has every line" without saying when it was last looked at
    // invites exactly the wrong conclusion.
    const when = row.age_days === null || row.age_days === undefined
      ? (row.state === 'never-captured' ? 'no capture stored' : '')
      : row.age_days < 1 ? 'captured today'
      : `captured ${Math.round(row.age_days)} day`
        + `${Math.round(row.age_days) === 1 ? '' : 's'} ago`;
    if (when) {
      head.appendChild(el('span', {
        class: row.stale ? 'compliance-age compliance-age-stale'
                         : 'compliance-age',
        text: row.stale ? `${when} — this verdict is that old too` : when,
      }));
    }
    card.appendChild(head);

    if (row.missing && row.missing.length) {
      card.appendChild(lineList('Missing', row.missing, 'compliance-missing'));
    }
    if (row.unexpected && row.unexpected.length) {
      card.appendChild(lineList('Should not be there', row.unexpected,
                                'compliance-unexpected'));
    }
    if (row.state === 'never-captured') {
      card.appendChild(el('p', { class: 'compliance-note',
        text: 'Nothing is claimed about this device either way. Back the '
            + 'group up, or open it once, and it will be answerable.' }));
    }
    if (row.state === 'no-snippet') {
      card.appendChild(el('p', { class: 'compliance-note',
        text: `No block was given for ${row.platform || 'this platform'}, so `
            + 'it was not checked. Checking it against another platform’s '
            + 'block would report every line missing.' }));
    }
    if (row.why) {
      card.appendChild(el('p', { class: 'compliance-note', text: row.why }));
    }

    // Only where there is something to fix and somewhere to fix it.
    if (row.state === 'missing' && row.missing && row.missing.length) {
      card.appendChild(el('button', {
        type: 'button', class: 'btn-secondary compliance-fix',
        title: 'Open this device and load the missing lines into the '
             + 'configuration editor, with a preview before anything is sent',
        onclick: () => openAndFix(row),
      }, [icon('tune'), 'Open and fix']));
    }
    return card;
  }

  function icon(name) {
    return el('span', { class: 'material-symbols-outlined', text: name });
  }

  function lineList(label, lines, cls) {
    return el('div', { class: `compliance-lines ${cls}` }, [
      el('span', { class: 'compliance-lines-label', text: label }),
      el('ul', {}, lines.map(line => el('li', {}, el('code', { text: line })))),
    ]);
  }

  /**
   * Open the device and load the missing lines into the push editor.
   *
   * Loaded, never sent. config_push shows its own preview and its own
   * dangerous-command guardrail, and this hands it text exactly as if
   * somebody had typed it — a compliance report is evidence, not
   * permission.
   */
  async function openAndFix(row) {
    const text = (row.missing || []).join('\n');
    const tab = await findOrOpen(row);
    if (!tab) {
      if (window.shellmateAlerts) window.shellmateAlerts.notify({
        global: true, severity: 'warning', icon: 'error',
        title: `Could not open ${row.name}`,
        body: 'Open a session to it, then use Apply configuration… from the '
            + 'tab menu. The missing lines are listed above.' });
      return;
    }
    close();
    if (window.shellmateConfigPush) window.shellmateConfigPush.open(tab, text);
  }

  /** An open tab for this device, or a new one from its saved profile. */
  async function findOrOpen(row) {
    const tabs = (typeof window.getOpenTabs === 'function')
      ? window.getOpenTabs() : [];
    const match = tabs.find(t =>
      String(t.hostname || '').toLowerCase() === String(row.hostname).toLowerCase()
      || String(t.label || '').toLowerCase() === String(row.name).toLowerCase());
    if (match) {
      // switchToTabBySessionId, not activateTab — the latter does not
      // exist. tabs.js exposes the session-id form for exactly this.
      if (typeof window.switchToTabBySessionId === 'function') {
        window.switchToTabBySessionId(match.sessionId);
      }
      return match;
    }

    if (typeof window.connectProfile !== 'function') return null;
    try {
      const res = await fetch('/api/profiles');
      const data = await res.json();
      const profiles = Array.isArray(data) ? data : (data.profiles || []);
      const profile = profiles.find(p => p.name === row.name
        || String(p.hostname || '').toLowerCase()
           === String(row.hostname).toLowerCase());
      if (!profile) return null;
      await window.connectProfile(profile);
    } catch (_) {
      return null;
    }

    // The tab appears asynchronously; give it a moment rather than racing.
    for (let attempt = 0; attempt < 20; attempt++) {
      await new Promise(r => setTimeout(r, 150));
      const tabsNow = (typeof window.getOpenTabs === 'function')
        ? window.getOpenTabs() : [];
      const found = tabsNow.find(t =>
        String(t.hostname || '').toLowerCase()
          === String(row.hostname).toLowerCase()
        || String(t.label || '').toLowerCase() === String(row.name).toLowerCase());
      if (found) return found;
    }
    return null;
  }

  document.addEventListener('DOMContentLoaded', init);

  window.shellmateCompliance = { open, run, show, close };
})();
